File: src/main/java/kaalka/KaalkaCLI.java
package kaalka;

import java.io.*;
import java.nio.file.*;
import java.util.*;

/**
 * KaalkaCLI: Command-line interface for Kaalka protocol
 * Supports envelope, text, and file encryption/decryption
 */
public class KaalkaCLI {
    public static void main(String[] args) {
        if (args.length == 0 || Arrays.asList(args).contains("--help")) {
            printUsage();
            return;
        }
        try {
            switch (args[0]) {
                case "encrypt":
                    handleEncrypt(args);
                    break;
                case "decrypt":
                    handleDecrypt(args);
                    break;
                case "envelope":
                    handleEnvelope(args);
                    break;
                default:
                    System.out.println("Unknown command. Use --help for usage.");
            }
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void printUsage() {
        System.out.println("""
Kaalka CLI (Java)
Usage:
  encrypt --in <inputFile> --out <outputFile> --sender <id> --receiver <id> --timestamp <time>
  decrypt --in <inputFile> --out <outputFile> --receiver <id> --timestamp <time>
  envelope --text "message" --sender <id> --receiver <id>
""");
    }

    private static void handleEncrypt(String[] args) throws Exception {
        String inFile = getArg(args, "--in");
        String outFile = getArg(args, "--out");
        String sender = getArg(args, "--sender");
        String receiver = getArg(args, "--receiver");
        String timestamp = getArg(args, "--timestamp");
        if (inFile == null || outFile == null || sender == null || receiver == null) {
            System.out.println("Missing required argument for encrypt");
            printUsage();
            return;
        }
        byte[] fileBytes = Files.readAllBytes(Paths.get(inFile));
        List<KaalkaEnvelope> chunks = KaalkaFile.encryptFileChunks(fileBytes, sender, receiver, timestamp);
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(outFile))) {
            oos.writeObject(chunks);
        }
        System.out.println("Chunks written to " + outFile);
    }

    private static void handleDecrypt(String[] args) throws Exception {
        String inFile = getArg(args, "--in");
        String outFile = getArg(args, "--out");
        String receiver = getArg(args, "--receiver");
        String timestamp = getArg(args, "--timestamp");
        if (inFile == null || outFile == null || receiver == null) {
            System.out.println("Missing required argument for decrypt");
            printUsage();
            return;
        }
        List<KaalkaEnvelope> chunks;
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(inFile))) {
            chunks = (List<KaalkaEnvelope>) ois.readObject();
        }
        byte[] fileBytes = KaalkaFile.decryptFileChunks(chunks, receiver, timestamp);
        Files.write(Paths.get(outFile), fileBytes);
        System.out.println("File written to " + outFile);
    }

    private static void handleEnvelope(String[] args) throws Exception {
        String text = getArg(args, "--text");
        String sender = getArg(args, "--sender");
        String receiver = getArg(args, "--receiver");
        String timestamp = getArg(args, "--timestamp");
        if (text == null || sender == null || receiver == null) {
            System.out.println("Missing required argument for envelope");
            printUsage();
            return;
        }
        KaalkaEnvelope env = KaalkaProtocol.encryptEnvelope(text, sender, receiver, timestamp, 120, 1);
        System.out.println("Envelope: " + env);
    }

    private static String getArg(String[] args, String name) {
        for (int i = 0; i < args.length - 1; i++) {
            if (args[i].equals(name)) return args[i + 1];
        }
        return null;
    }
}

File: src/main/kotlin/kaalka/KaalkaCLI.kt
package kaalka

import java.io.File
import kotlin.system.exitProcess

/**
 * Kaalka CLI tool
 * Supports envelope, text, and file encryption/decryption
 */
object KaalkaCLI {
    @JvmStatic
    fun main(args: Array<String>) {
        if (args.isEmpty() || args.contains("--help")) {
            printUsage()
            exitProcess(0)
        }
        try {
            when (args[0]) {
                "encrypt" -> handleEncrypt(args)
                "decrypt" -> handleDecrypt(args)
                "envelope" -> handleEnvelope(args)
                else -> {
                    println("Unknown command. Use --help for usage.")
                    exitProcess(1)
                }
            }
        } catch (e: Exception) {
            println("Error: ${e.message}")
            exitProcess(1)
        }
    }

    private fun printUsage() {
        println("""
Kaalka CLI (Kotlin)
Usage:
  encrypt --in <inputFile> --out <outputFile> --sender <id> --receiver <id> --timestamp <time>
  decrypt --in <inputFile> --out <outputFile> --receiver <id> --timestamp <time>
  envelope --text "message" --sender <id> --receiver <id>
""")
    }

    private fun handleEncrypt(args: Array<String>) {
        val inFile = getArg(args, "--in")
        val outFile = getArg(args, "--out")
        val sender = getArg(args, "--sender")
        val receiver = getArg(args, "--receiver")
        val timestamp = getArg(args, "--timestamp")
        if (inFile == null || outFile == null || sender == null || receiver == null) {
            println("Missing required argument for encrypt")
            printUsage()
            exitProcess(1)
        }
        val fileBytes = File(inFile).readBytes()
        val chunks = KaalkaFile.encryptFileChunks(fileBytes, sender, receiver, timestamp)
        File(outFile).writeBytes(chunks.joinToString("\n") { it.toString() }.toByteArray())
        println("Chunks written to $outFile")
    }

    private fun handleDecrypt(args: Array<String>) {
        val inFile = getArg(args, "--in")
        val outFile = getArg(args, "--out")
        val receiver = getArg(args, "--receiver")
        val timestamp = getArg(args, "--timestamp")
        if (inFile == null || outFile == null || receiver == null) {
            println("Missing required argument for decrypt")
            printUsage()
            exitProcess(1)
        }
        val lines = File(inFile).readLines()
        val chunks = lines.map { parseEnvelope(it) }
        val fileBytes = KaalkaFile.decryptFileChunks(chunks, receiver, timestamp)
        File(outFile).writeBytes(fileBytes)
        println("File written to $outFile")
    }

    private fun handleEnvelope(args: Array<String>) {
        val text = getArg(args, "--text")
        val sender = getArg(args, "--sender")
        val receiver = getArg(args, "--receiver")
        val timestamp = getArg(args, "--timestamp")
        if (text == null || sender == null || receiver == null) {
            println("Missing required argument for envelope")
            printUsage()
            exitProcess(1)
        }
        val env = KaalkaProtocol.encryptEnvelope(text, sender, receiver, timestamp)
        println("Envelope: $env")
    }

    private fun getArg(args: Array<String>, name: String): String? {
        val idx = args.indexOf(name)
        return if (idx != -1 && idx + 1 < args.size) args[idx + 1] else null
    }

    private fun parseEnvelope(str: String): KaalkaEnvelope {
        // Simple parser for KaalkaEnvelope from string (for demo purposes)
        val regex = Regex("KaalkaEnvelope\\{senderId='(.*?)', receiverId='(.*?)', timestamp='(.*?)', window=(\\d+), seq=(\\d+), ciphertext='(.*?)', seal='(.*?)'\\}")
        val match = regex.matchEntire(str)
        if (match != null) {
            val (senderId, receiverId, timestamp, window, seq, ciphertext, seal) = match.destructured
            return KaalkaEnvelope(senderId, receiverId, timestamp, window.toInt(), seq.toInt(), ciphertext, seal)
        }
        throw Exception("Invalid envelope format")
    }
}
